### PR TITLE
Multi process support

### PR DESCRIPTION
**Pull request by simonvadee**
J'ai un doute sur la classe Process (src/process.js) : étant donné que c'est formaté en classe, les méthodes qui auparavant n'étaient pas accessibles directement depuis l'agent (fossilizeSegment, saveSegment, applyPlugins) le sont maintenant depuis une instance de la classe Process. Pour reproduire le comportement d'une méthode privée, il faudrait instancier les fonctions dans le constructeur (est-ce la seule manière ?), ce que je ne me suis pas résolu à faire. C'est un détail mais si vous avez une solution au problème, je suis preneur !

### CODE REVIEW

File: src/create.js [hunk 1]
```diff
@@ -14,266 +14,218 @@
...
+
+      for (const process of filteredProcesses) {
+        results.push(process.getMapIds(opts));
+      }
+      return Promise.all(results).then(res => {
+        if (res.length) {
+          return Promise.resolve(res.reduce((acc, val) => [...acc, ...val]));
+        }
+        return Promise.resolve([]);
+      });
```
Comment (such) [inline src/create.js]: We could use the same idea than before

File: src/httpServer.js [hunk 2]
```diff
@@ -22,13 +22,26 @@ import parseArgs from './parseArgs';
 
 /**
  * Creates an HTTP server for an agent.
- * @param {Agent} agent - the agent instance
+ * @param {object} processes - the array of processes to mount as subpath
```
Comment (such) [inline src/httpServer.js]: we're still passing the agent right?

File: src/httpServer.js [hunk 3]
```diff
@@ -22,13 +22,26 @@ import parseArgs from './parseArgs';
...
  * @param {object} [opts] - options
  * @param {object} [opts.cors] - CORS options
  * @param {object} [opts.salt] - salt used for callback URLs
  * @returns {express.Server} an express server
  */
 export default function httpServer(agent, opts = {}) {
+  function getProcessByName(name) {
+    return agent.getAllProcesses()
```
Comment (such) [inline src/httpServer.js]: we could delegate the call completely to the agent, that would be more efficient I think.

File: src/process.js [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+  constructor(name, actions, storeClient, fossilizerClient, opts = {}) {
+    this.name = name;
+    this.actions = actions;
+    this.storeClient = storeClient;
+    this.fossilizerClient = fossilizerClient;
+    this.salt = opts.salt || '';
+    this.plugins = opts.plugins || [];
+    this.evidenceCallbackUrl = opts.evidenceCallbackUrl || '';
+    this.agentUrl = opts.agentUrl || '';
+    this.agentInfo = {
```
Comment (simonvadee) [inline src/process.js]: This member should be renamed processInfo

File: src/httpServer.js [hunk 5]
```diff
@@ -170,5 +246,7 @@ export default function httpServer(agent, opts = {}) {
 
   app.use(error());
 
-  return app;
+  return {
+    app,
```
Comment (such) [inline src/httpServer.js]: why do we want to return an object?

**Comment by such** [comment]
English please ;)

**Comment by such** [comment]
I think we can ignore the function access control for now. @stephan83 what do you think?

File: src/create.js [hunk 6]
```diff
@@ -14,266 +14,218 @@
...
+              }
+              return {
+                process: name,
+                actions: getActionsInfo(processes[name].actions),
+                pluginsInfo: getPluginsInfo(processes[name].plugins),
+                storeInfo: _storeInfo
+              };
+            });
+      }
+      return Promise.all(Object.values(agentInfo))
```
Comment (simonvadee) [inline src/create.js]: This should be reformatted as : {processesInfo: Array<Object>}

**Comment by such** [comment]
I have a feeling the commits are not that atomic. It's no big deal but in that case it would probably be better to rebase everything down to one commit

File: src/create.js (hunk 6 above)
Comment (such) [inline src/create.js]: I like having the info indexed by processName

File: src/create.js [hunk 7]
```diff
@@ -14,266 +14,218 @@
...
      * @param {string} [opts.prevLinkHash] - a previous link hash the segments must have
      * @param {string[]} [opts.tags] - an array of tags the segments must have
      * @returns {Promise} a promise that resolve with the segments
      */
-    findSegments(options) {
-      return storeClient.findSegments(options, getDefinedFilters(plugins));
+    findSegments(opts = {}) {
+      const results = [];
+      const filteredProcesses = opts.processesName ?
+        Object.values(processes).filter(p => opts.processesName.indexOf(p.name) >= 0)
```
Comment (simonvadee) [inline src/create.js]: but it doesn't check wether `processes[name]` exists, and fills the list with an 'undefined' if it doesn't. Anyways I agree that it's not very good, I'll look into it ;)

**Comment by stephan83** [comment]
Pour les "classes" comme vous voulez, il y a les deux écoles... Mais quand vous avez fait votre choix ajoutez le dans les guidelines Javascript :)

File: src/create.js (hunk 7 above)
Comment (such) [inline src/create.js]: you could do:
`opts.processesName.map(name => processes[name]).filter(p => p)`

File: src/create.js [hunk 8]
```diff
@@ -14,266 +14,218 @@
...
-
-  // Connect to store web socket.
-  storeClient.connect(opts.reconnectTimeout || 5000);
-
-  function applyPlugins(method, ...args) {
-    // execute all plugins sequentially
-    return plugins.reduce(
-      (cur, plugin) => cur.then(() => Promise.resolve(plugin[method] && plugin[method](...args))),
-      Promise.resolve());
+    return Promise.resolve(p);
```
Comment (simonvadee) [inline src/create.js]: we don't

File: src/create.js [hunk 9]
```diff
@@ -14,266 +14,218 @@
...
-    return segment;
-  }
+export default function create(options) {
+  const processes = Object();
+  const storeClients = [];
+  let server;
+
+  function initStoreClient(storeClient, reconnectTimeout = 5000) {
+    storeClients.push(storeClient);
+    // TODO : check if callbacks are called when several processes are run
```
Comment (simonvadee) [inline src/create.js]: it wasn't, it will be in the next push.

File: src/create.js [hunk 10]
```diff
@@ -14,266 +14,218 @@
...
+      for (const process of filteredProcesses) {
+        results.push(process.findSegments(opts));
+      }
+      // waits for all the segments and concatenate them into one array.
+      return Promise.all(results).then(res => {
+        if (res.length) {
+          return Promise.resolve(res.reduce((acc, val) => [...acc, ...val]));
+        }
+        return Promise.resolve([]);
+      });
```
Comment (simonvadee) [inline src/create.js]: much better indeed

File: src/create.js [hunk 11]
```diff
@@ -14,266 +14,211 @@
...
-
-          segment = { link, meta };
-
-          return applyPlugins('didCreateSegment', segment, 'init', args);
-        })
-        .then(
-          () => saveSegment(segment)
-        );
+    addProcess(processName, actions, storeClient, fossilizerClient, opts = {}) {
+      if (typeof processes[processName] === 'object') {
```
Comment (such) [inline src/create.js]: isn't it enough to check `if (processes[processName])`?

File: src/httpServer.js [hunk 12]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
+      .getMapIds(req.query)
+      .then(res.json.bind(res))
+      .catch(next);
+  });
+
+  app.get('/processes', (req, res, next) => {
+    Promise.resolve(agent
+      .getAllProcesses(req.query))
+      .then(res.json.bind(res))
+      .catch(next);
```
Comment (such) [inline src/httpServer.js]: I'm not such a fan of forcing everything into a Promise. Would this work?
```
const processes = agent.getAllProcesses(req.query)
res.json(processes)
```

File: src/httpServer.js [hunk 13]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
+  });
+
+  app.get('/:process/remove', (req, res, next) => {
+    try {
+      const processList = agent.removeProcess(req.params.process);
+      Promise.resolve(processList).then(res.json.bind(res));
```
Comment (such) [inline src/httpServer.js]: Same as before:
```
res.json(processList)
```
should be enough?

File: src/httpServer.js [hunk 14]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
     /*eslint-enable*/
 
-    agent
-      .createMap(...parseArgs(req.body))
-      .then(res.json.bind(res))
+    Promise.resolve(agent.getProcess(req.params.process))
+      .then(process =>
+        process.createMap(...parseArgs(req.body))
+          .then(res.json.bind(res))
+          .catch(next))
```
Comment (such) [inline src/httpServer.js]: I would also split it into two statements:
```
const process = agent.getProcess(req.params.process);
process.createMap(...parseArgs(req.body))
     .then(res.json.bind(res))
    .catch(next));
```

File: src/httpServer.js [hunk 15]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
 
-    agent
-      .createSegment(req.params.linkHash, req.params.action, ...parseArgs(req.body))
-      .then(res.json.bind(res))
+    Promise.resolve(agent.getProcess(req.params.process)).then(process => {
+      process
+        .createSegment(req.params.linkHash, req.params.action, ...parseArgs(req.body))
+        .then(res.json.bind(res))
+        .catch(next);
+    })
```
Comment (such) [inline src/httpServer.js]: See above

File: src/httpServer.js [hunk 16]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
-    agent
-      .getSegment(req.params.linkHash)
-      .then(res.json.bind(res))
+  app.get('/:process/segments/:linkHash', (req, res, next) => {
+    Promise.resolve(agent.getProcess(req.params.process)).then(process => {
+      process
+        .getSegment(req.params.linkHash)
+        .then(res.json.bind(res))
+        .catch(next);
+    })
```
Comment (such) [inline src/httpServer.js]: See above

File: src/httpServer.js [hunk 17]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
-    agent
-      .findSegments(req.query)
-      .then(res.json.bind(res))
+  app.get('/:process/segments', (req, res, next) => {
+    Promise.resolve(agent.getProcess(req.params.process)).then(process => {
+      process
+        .findSegments(req.query)
+        .then(res.json.bind(res))
+        .catch(next);
+    })
```
Comment (such) [inline src/httpServer.js]: See above

File: src/httpServer.js [hunk 18]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
-    agent
-      .insertEvidence(req.params.linkHash, req.body, req.query.secret)
-      .then(res.json.bind(res))
+  app.post('/:process/evidence/:linkHash', (req, res, next) => {
+    Promise.resolve(agent.getProcess(req.params.process)).then(process => {
+      process
+        .insertEvidence(req.params.linkHash, req.body, req.query.secret)
+        .then(res.json.bind(res))
+        .catch(next);
+    })
```
Comment (such) [inline src/httpServer.js]: See above

File: src/httpServer.js [hunk 19]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
-    agent
-      .getMapIds(req.query)
-      .then(res.json.bind(res))
+  app.get('/:process/maps', (req, res, next) => {
+    Promise.resolve(agent.getProcess(req.params.process)).then(process => {
+      process
+        .getMapIds(req.query)
+        .then(res.json.bind(res))
+        .catch(next);
+    })
```
Comment (such) [inline src/httpServer.js]: See above

File: src/httpServer.js [hunk 20]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
 
-    agent
-      .createMap(...parseArgs(req.body))
-      .then(res.json.bind(res))
+    Promise.resolve(agent.getProcess(req.params.process)).then(process => {
+      process
+        .createMap(...parseArgs(req.body))
+        .then(res.json.bind(res))
+        .catch(next);
+    })
```
Comment (such) [inline src/httpServer.js]: see above

File: src/httpServer.js [hunk 21]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
 
-    agent
-      .createSegment(req.params.linkHash, req.params.action, ...parseArgs(req.body))
-      .then(res.json.bind(res))
+    Promise.resolve(agent.getProcess(req.params.process)).then(process => {
+      process
+        .createSegment(req.params.linkHash, req.params.action, ...parseArgs(req.body))
+        .then(res.json.bind(res))
+        .catch(next);
+    })
```
Comment (such) [inline src/httpServer.js]: see above

File: src/httpServer.js [hunk 22]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
-    agent
-      .getSegment(req.params.linkHash)
-      .then(res.json.bind(res))
+  app.get('/:process/links/:linkHash', (req, res, next) => {
+    Promise.resolve(agent.getProcess(req.params.process)).then(process => {
+      process
+        .getSegment(req.params.linkHash)
+        .then(res.json.bind(res))
+        .catch(next);
+    })
```
Comment (such) [inline src/httpServer.js]: see above

File: src/httpServer.js [hunk 23]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
-    agent
-      .findSegments(req.query)
-      .then(res.json.bind(res))
+  app.get('/:process/links', (req, res, next) => {
+    Promise.resolve(agent.getProcess(req.params.process)).then(process => {
+      process
+        .findSegments(req.query)
+        .then(res.json.bind(res))
+        .catch(next);
+    })
```
Comment (such) [inline src/httpServer.js]: see above

File: src/httpServer.js [hunk 24]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
 
-    agent
-      .findSegments(req.query)
-      .then(res.json.bind(res))
+    Promise.resolve(agent.getProcess(req.params.process)).then(process => {
+      process
+        .findSegments(req.query)
+        .then(res.json.bind(res))
+        .catch(next);
+    })
```
Comment (such) [inline src/httpServer.js]: see above

File: src/httpServer.js [hunk 25]
```diff
@@ -48,117 +48,182 @@ export default function httpServer(agent, opts = {}) {
...
```
Comment (such) [inline src/httpServer.js]: see above

File: src/process.js [hunk 26]
```diff
@@ -0,0 +1,255 @@
...
+  applyPlugins(method, ...args) {
+    // execute all plugins sequentially
+    return this.plugins.reduce(
+      (cur, plugin) => cur.then(() => Promise.resolve(plugin[method] && plugin[method](...args))),
+      Promise.resolve());
+  }
+
+  /**
+  * Gets information about the agent.
+  * @returns {Promise} - wxa promise that resolve with the information
```
Comment (such) [inline src/process.js]: typos :)
Gets information about the process actually as well

File: src/create.js [hunk 27]
```diff
@@ -14,266 +14,211 @@
...
+              }
+              return {
+                process: name,
+                actions: getActionsInfo(processes[name].actions),
+                pluginsInfo: getPluginsInfo(processes[name].plugins),
+                storeInfo: _storeInfo
+              };
+            });
+        return acc;
+      }, {});
```
Comment (such) [inline src/create.js]: why don't we use directly process.getInfo() here?

File: src/create.js [hunk 28]
```diff
@@ -14,266 +14,211 @@
...
-    findSegments(options) {
-      return storeClient.findSegments(options, getDefinedFilters(plugins));
+    findSegments(opts = {}) {
+      const filteredProcesses = opts.processes ?
+        opts.processes.map(name => processes[name]).filter(p => p)
+        : Object.values(processes);
+
+      // waits for all the segments and concatenate them into one array.
+      return Promise.all(filteredProcesses.map(p => p.findSegments(opts)))
+        .then(all => all.reduce((acc, cur) => acc.concat(cur), []));
```
Comment (such) [inline src/create.js]: As seen together we should handle better the case when multiple processes use the same store client.

File: src/create.js (hunk 27 above)
Comment (simonvadee) [inline src/create.js]: good point...

File: src/process.js [hunk 29]
```diff
@@ -60,6 +60,7 @@ export default class Process {
   }
 
   saveSegment(segment) {
+    console.log("SAVING == ", segment)
```
Comment (such) [inline src/process.js]: Is this log really helpful?

**Comment by simonvadee** [comment]
Done rebasing and editing commit message, is it looking good enough to merge ?

**Comment by such** [comment]
👍

**Pull request merged**